### ISSUE TITLE
mel_layer可能存在问题

### ISSUE THREAD

**Issue by phecda-xu** [Z-yq/TensorflowASR#13]
你好：

基于你给的框架我用一维膨胀卷积（dilate Conv1D）搭建了一个TDNNTransducer的网络，尝试训练时遇到下面这种情况的报错：

`indices[1] = [1 , 776, 22] does not index into shape [4, 776, 23]`

错误位置在
[这里](https://github.com/Z-yq/TensorflowASR/blob/128dac733b9c34ca097ce6406b229f63fe157c1e/losses/rnnt_losses.py#L175)
`last_grads_blank = -1 * tf.scatter_nd(
        tf.concat([tf.reshape(tf.range(batch_size, dtype=tf.int32),
                              shape=[batch_size, 1]), indices], axis=1),
        tf.ones(batch_size, dtype=tf.float32), [batch_size, input_max_len, target_max_len])`

打印中间的信息：

```
input_max_len = 776

logit_length = [398, 777, 378, 777]

indices = 
[[0 397 12]
 [1 776 22]
 [2 377 12]
 [3 776 22]]

[batch_size, input_max_len, target_max_len]  = [4, 776, 23]
```
tf.scatter_nd() 使用时要求indices 每一列的最大值要比 [batch_size, input_max_len, target_max_len]  对应的值小，否则就会报出前面的索引错误。

至于为什么会出现相等的情况（示例中indices是776，与input_max_len值相等），从logit_length 可以看出，网络计算的logits对应维度大小为777，而给定的是776，实际上logits对应维度应该是776，计算indices时 776-1 = 775 ，这样才能满足条件。

[这里](https://github.com/Z-yq/TensorflowASR/blob/128dac733b9c34ca097ce6406b229f63fe157c1e/trainer/transducer_runners.py#L46)可以看到，input_max_len是生成batch时，用传统方法提取声学特征计算出来的，当使用mel_layer时，是直接输入的raw数据，也就是说用传统方法提取声学特征和mel_layer自动处理raw数据生成的特征帧数是不一致的，**mel_layer多了一帧**。

将 use_mel_layer 设置为False后，训练正常了，验证了我的判断。

另外，这个错误不是立刻出现的，是经过十几个batch的正常计算后出现的。

**Comment by Z-yq** [comment]
这个问题是必出现的吗？还是偶尔某几个数据的问题？

> indices[1] = [1 , 776, 22] does not index into shape [4, 776, 23]
从这里看，更像是网络的输出不匹配原始数据的batch了

**Comment by phecda-xu** [comment]
> 这个问题是必出现的吗？还是偶尔某几个数据的问题？
> 
> > indices[1] = [1 , 776, 22] does not index into shape [4, 776, 23]
> > 从这里看，更像是网络的输出不匹配原始数据的batch了

我觉得是必然出现的，因为试过多次，每次都会报错，且每次出问题时的数据形状都不一样，不可能是偶尔的某几个数据（同样的数据在训练你提供的网络是没问题的）。

至于下面的你说看上去网络的输出不必配原始数据的batch，实际上是 `tf.scatter_nd`这个函数的的内部计算过程，
先生成 indices = 
[[0 397 12]
 [1 776 22]
 [2 377 12]
 [3 776 22]]

而 [1 , 776, 22] 是indices其中的一行，然后 `tf.scatter_nd`执行时会在一个tensor的内部按照`[1 , 776, 22]` 的索引去插入0值，实际上应该是775（776 - 1），但是生成的是776（777 -1），超界限了。（用mel_layer是777-1，不用是776-1）

具体rnnt-loss的计算过程我还没看懂，不明白这个一步的目的，但是可以确信错误是在这里出现的。

另外，我怀疑，没有帧压缩的网络应该都会出现这种情况（比如纯RNN类的），我的TDNN，输入N帧的声学特征输出的是N帧的高维特征，帧数上没变，用mel_layer后恰好就出界了（只出界1帧）。而其它网络结构（CRNN/CNN等）出现帧数压缩的网络，大概不会出现这个问题，因为高维特征的帧数肯定低于声学特征帧数（此时mel_layer多1帧不影响）。

**Comment by Z-yq** [comment]
model 里的self.time_reduction_factor  有重新设置吗？
并且和am_data.yml里保持一致?

**Comment by phecda-xu** [comment]
> model 里的self.time_reduction_factor 有重新设置吗？
> 并且和am_data.yml里保持一致?

设置了，都为1，实际上网络内部没用到这参数。
我在fork的仓库里上传了 我的网络和配置[tdnn](https://github.com/phecda-xu/TensorflowASR/commit/8a9d8152371b7e9d3354c60fcb8f3b4d4328ca47)，你可以试试看。

**Comment by Z-yq** [comment]
这就非常有意思了，在我这里，报错是因为卷积的空洞操作引起的。取消空洞操作之后一切正常。
自己暂时没有时间去跟这个问题。

**Comment by phecda-xu** [comment]
> 这就非常有意思了，在我这里，报错是因为卷积的空洞操作引起的。取消空洞操作之后一切正常。
> 自己暂时没有时间去跟这个问题。

哦对，忘了说，要tensorflow2.3以上才可以，2.2以及以下的版本空洞卷积是有问题的。先这样吧，你有时间再说，谢谢了！
